src/db/package_version.rs

```diff
@@ -0,0 +1,168 @@
+    ) -> Result<models::PackageVersion, DatabaseError> {
+        // Check if the package exists.
+        let pkg_name = request.package_name.clone();
+        let package = if let Some(existing_package) = schema::packages::table
+            .filter(schema::packages::package_name.eq(pkg_name.clone()))
+            .select(schema::packages::all_columns)
+            .first::<models::Package>(self.inner())
+            .optional()
+            .map_err(|err| DatabaseError::QueryFailed(pkg_name.clone(), err))?
+        {
+            if existing_package.user_owner != api_token.user_id {
+                // The package exists but the user is not the owner.
+                return Err(DatabaseError::InvalidPublishToken);
+            }
+            Ok(existing_package)
+        } else {
+            // Insert a new package.
+            let new_package: models::NewPackage = models::NewPackage {
+                user_owner: api_token.user_id,
+                package_name: pkg_name.clone(),
+            };
+
+            let saved_package = diesel::insert_into(schema::packages::table)
+                .values(&new_package)
+                .returning(models::Package::as_returning())
+                .get_result(self.inner())
+                .map_err(|err| DatabaseError::InsertPackageFailed(pkg_name.clone(), err))?;
+            Ok(saved_package)
+        }?;
+
+        // Insert a new package version.
+        let new_version = models::NewPackageVersion {
+            package_id: package.id,
+            publish_token: api_token.id,
+            published_by: api_token.user_id,
+            upload_id: request.upload_id,
+            num: request.num.clone(),
+            package_description: request.package_description.clone(),
+            repository: request.repository.clone(),
+            documentation: request.documentation.clone(),
+            homepage: request.homepage.clone(),
+            urls: request.urls.clone(),
+            readme: request.readme.clone(),
+            license: request.license.clone(),
+        };
+
+        let saved_version = diesel::insert_into(schema::package_versions::table)
+            .values(&new_version)
+            .returning(models::PackageVersion::as_returning())
+            .get_result(self.inner())
+            .map_err(|err| {
+                DatabaseError::InsertPackageVersionFailed(
+                    pkg_name.clone(),
+                    request.num.clone(),
+                    err,
+                )
+            })?;
+
+        // Update the package default version.
+        diesel::update(schema::packages::table.filter(schema::packages::id.eq(package.id)))
+            .set(schema::packages::default_version.eq(saved_version.id))
+            .execute(self.inner())
+            .map_err(|err| DatabaseError::UpdatePackageFailed(pkg_name.clone(), err))?;
+
+        Ok(saved_version)
+    }
+
+    /// Fetch a package given the package ID.
+    pub fn get_package_by_id(
+        &mut self,
```

Comment:
wondering why this function and the next param needs to be `mut` since we dont seem to be making any state changes, Not sure whether diesel requires mutable

Reply:
diesel needs a mutable reference to the `DBConn` to make queries, whether they are mutations or not.